langchain/callbacks/mlflow_callback.py

```diff
@@ -0,0 +1,691 @@
+        }
+
+        resp.update(text_visualizations)
+
+    return resp
+
+
+def construct_html_from_prompt_and_generation(prompt: str, generation: str) -> Any:
+    """Construct an html element from a prompt and a generation.
+
+    Parameters:
+        prompt (str): The prompt.
+        generation (str): The generation.
+
+    Returns:
+        (str): The html string."""
+    formatted_prompt = prompt.replace("\n", "<br>")
+    formatted_generation = generation.replace("\n", "<br>")
+
+    return f"""
+    <p style="color:black;">{formatted_prompt}:</p>
+    <blockquote>
+      <p style="color:green;">
+        {formatted_generation}
+      </p>
+    </blockquote>
+    """
+
+
+class MlflowLogger:
+    """Callback Handler that logs metrics and artifacts to mlflow server.
+
+    Parameters:
+        name (str): Name of the run.
+        experiment (str): Name of the experiment.
+        tags (str): Tags to be attached for the run.
+        tracking_uri (str): MLflow tracking server uri.
+
+    This handler implements the helper functions to initialize,
+    log metrics and artifacts to the mlflow server.
+    """
+
+    def __init__(self, **kwargs):
+        mlflow = import_mlflow()
+        tracking_uri = get_from_dict_or_env(
+            kwargs, "tracking_uri", "MLFLOW_TRACKING_URI"
+        )
+        self.mlfc = mlflow.MlflowClient(tracking_uri=tracking_uri)
+
+        # User can set other env variables described here
+        # > https://www.mlflow.org/docs/latest/tracking.html#logging-to-a-tracking-server
+
+        experiment_name = get_from_dict_or_env(
+            kwargs, "experiment_name", "MLFLOW_EXPERIMENT_NAME"
+        )
+        self.mlf_exp = self.mlfc.get_experiment_by_name(experiment_name)
+        if self.mlf_exp is not None:
+            self.mlf_exp_id = self.mlf_exp.experiment_id
+        else:
+            self.mlf_exp_id = self.mlfc.create_experiment(experiment_name)
+            self.mlf_exp = self.mlfc.get_experiment_by_name(experiment_name)
+
+        self.start_run(kwargs["run_name"], kwargs["run_tags"])
+
+    def start_run(self, name: str, tags: list = []):
+        """To start a new run, auto generates the random suffix for name"""
+        if name.endswith("-%"):
+            rname = "".join(random.choices(string.ascii_uppercase + string.digits, k=7))
+            name = name.replace("%", rname)
+        self.run = self.mlfc.create_run(self.mlf_exp_id, run_name=name)
+
+    def finish_run(self):
+        """To finish the run."""
+        self.mlfc.set_terminated(self.run.info.run_id)
+
+    def metric(self, key: str, value: float):
+        """To log metric to mlflow server."""
+        # metric value only of type float or int
+        if not isinstance(value, float) and not isinstance(value, int):
+            return
+        self.mlfc.log_metric(self.run.info.run_id, key, value)
+
+    def json(self, data: Any):
+        """To log all metrics in the input dict."""
+        for k, v in data.items():

```

Comment:
can we convert this to a log_batch instead of repeated network requests in a loop?

Reply:
Done !